File: 3_longest_substring_without_repeating_characters.py
# beats 99%

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        return_ = 0

        for elem_index in range(0, len(s)):
            stack_string = []

            for letter in s[elem_index:]:
                
                if not letter in stack_string:
                    stack_string.append(letter)
            
                else:
                    break

            if len(stack_string) > return_:
                return_ = len(stack_string)

        return return_

print(Solution().lengthOfLongestSubstring("abcabcbb")) # -> 3
print(Solution().lengthOfLongestSubstring("bbbbb")) # -> 1
print(Solution().lengthOfLongestSubstring("pwwkew")) # -> 3

File: 3_longest_substring_without_repeating_characters.go
// https://leetcode.com/problems/longest-substring-without-repeating-characters/

package main

import (
	"fmt"
)

// beats 100%

func lengthOfLongestSubstring(s string) int {
	result := 0
	left := 0
	char_map := make(map[byte]int)

	for right := 0; right < len(s); right++ {
		if index, found := char_map[s[right]]; found && index >= left {
			left = index + 1
		}
		char_map[s[right]] = right

		current_len := right - left + 1
		if current_len > result {
			result = current_len
		}
	}

	return result
}

func main() {
	fmt.Println(lengthOfLongestSubstring("abcabcbb")) // 3
	fmt.Println(lengthOfLongestSubstring("bbbbb"))    // 1
	fmt.Println(lengthOfLongestSubstring("pwwkew"))   // 3
}
